### ISSUE TITLE
Delta update issue

### ISSUE THREAD

**Issue by zengxiyu** [Opdoop/GenerAT#2]
After adding print to both the if and else branches of the update_delta function, I found that delta only performs initialization operations and does not update based on gradients. Does this approach seem to contradict the statement in the paper?

**Comment by Opdoop** [comment]
@zengxiyu  Which function are you looking at? Can you give a code line reference?

**Comment by Opdoop** [comment]
Please provide a script to reproduce the results or the complete log to check your issues.

**Comment by zengxiyu** [comment]
In the update_delta method of the PerturbionLayer class in the https://github.com/Opdoop/GenerAT/blob/main/DeBERTa/sift/sift.py file, add print to the if and else branches for initializing delta and updating delta, and then run bash/ Adv_glue/rte.sh found that only the branch that initializes delta will print.

**Comment by Opdoop** [comment]
The forward attack calculation is in:
https://github.com/Opdoop/GenerAT/blob/86fee393359bf2d0ad37148f21b306f1ebd43b51/DeBERTa/sift/sift.py#L255-L272
In line 255 we calculate the adv loss of the encoder.
In line 257 we call the update_delta to change the forward behavior to get perturbed representation. As this line
https://github.com/Opdoop/GenerAT/blob/86fee393359bf2d0ad37148f21b306f1ebd43b51/DeBERTa/sift/sift.py#L58
In line 263 we get the generated adversarial example.
In your initial question, you say ` contradict the statement in the paper`.

What's your question exactly? Can you describe it step by step? e.g. what process contradicts to which implementation

**Comment by zengxiyu** [comment]
I just modified this function, as shown in the following code：
```python
def update_delta(self, requires_grad=False):
        if not self.adversarial_mode:
            return True
        if self.delta is None:
            print("delta initialization")
            delta = torch.clamp(self.input.new(self.input.size()).normal_(0, self.init_perturbation).float(),
                                -2 * self.init_perturbation, 2 * self.init_perturbation)
        else:
            print("delta is based on gradient updates")
            for line in traceback.format_stack():
                print(line.strip())
            grad = self.delta.grad
            self.delta.grad = None
            delta = self.delta
            norm = grad.norm()
            if torch.isnan(norm) or torch.isinf(norm):
                return False
            eps = self.learning_rate
            with torch.no_grad():
                delta = delta + eps * grad / (1e-6 + grad.abs().max(-1, keepdim=True)[0])
        self.delta = delta.float().detach().requires_grad_(requires_grad)
        self.perturbated_input = (self.input.to(delta).detach() + self.delta).to(self.input)
        return True
```
Later, when I ran the rte.sh script, I found that the program only prints "delta initialization".This means that delta has not performed any update operations.
My question is that you proposed in your paper to add **continuous perturbations** in the embedding layer, while in the program, it seems that only the **initialized delta is added to the embedding layer**，which does not reflect persistent perturbations.
And it's in the sift file
https://github.com/Opdoop/GenerAT/blob/86fee393359bf2d0ad37148f21b306f1ebd43b51/DeBERTa/sift/sift.py#L253-L267
GenerativeAdversarialLearner.loss always sets delta to None in the end, which I think is the reason for this.

**Comment by Opdoop** [comment]
@zengxiyu Thanks Zeng! You are right! This must be a bug! I'm trying to fix this.

**Comment by Opdoop** [comment]
I took it for granted that this if-else can update delta.
https://github.com/Opdoop/GenerAT/blob/86fee393359bf2d0ad37148f21b306f1ebd43b51/DeBERTa/sift/sift.py#L41-L59
You see when `self.delta is None` we use a random initialization. 
Next, in line 57 `self.delta = delta.float().detach().requires_grad_(requires_grad) `
So, in next iteration, I think it will enter the `else` branch to get the delta by grad. 
Unfortunately, as you pointed out, it never enters the `else` branch.

**Comment by Opdoop** [comment]
I think this is caused by the `cleanup` function in lien 266.
https://github.com/Opdoop/GenerAT/blob/86fee393359bf2d0ad37148f21b306f1ebd43b51/DeBERTa/sift/sift.py#L266

**Comment by Opdoop** [comment]
I modified the `update_delta` so that it always gets the grad of the embedding layer. 
https://github.com/Opdoop/GenerAT/blob/fff8166d1468131338ac48ba4bbf94c6cdac9166/DeBERTa/sift/sift.py#L42-L51
You can pull the new code and run it. If you have any questions, feel free to ask. Thanks again.

**Comment by zengxiyu** [comment]
Thank you for your response！
